src/v/kafka/server/handlers/fetch.cc

```diff
@@ -702,6 +711,186 @@ class parallel_fetch_plan_executor final : public fetch_plan_executor::impl {
     }
 };
 
+class pid_controller {
+    // 0.1234 * scaling_factor = 1234
+    static constexpr int64_t scaling_factor = 10000;
+    static constexpr auto duration_unit = 1us;
+    static constexpr auto sample_rate = 250us;
+
+public:
+    pid_controller()
+      : _last_sampled_time{ss::sched_clock::now()}
+      , _p_coeff{config::shard_local_cfg().fetch_pid_p_coeff()}
+      , _i_coeff{config::shard_local_cfg().fetch_pid_i_coeff()}
+      , _d_coeff{config::shard_local_cfg().fetch_pid_d_coeff()}
+      , _target_fetch_sg_util{(
+          int64_t)(scaling_factor
+                   * config::shard_local_cfg().fetch_pid_target_fetch_sg())}
+      , _max_delay{config::shard_local_cfg().fetch_pid_max_debounce()}
+      , _p_coeff_binding{config::shard_local_cfg().fetch_pid_p_coeff.bind()}
+      , _i_coeff_binding{config::shard_local_cfg().fetch_pid_i_coeff.bind()}
+      , _d_coeff_binding{config::shard_local_cfg().fetch_pid_d_coeff.bind()}
+      , _target_fetch_sg_util_binding{config::shard_local_cfg()
+                                        .fetch_pid_target_fetch_sg.bind()}
+      , _max_delay_binding{
+          config::shard_local_cfg().fetch_pid_max_debounce.bind()} {
+        _p_coeff_binding.watch([this] {
+            reset();
+            _p_coeff = config::shard_local_cfg().fetch_pid_p_coeff();
+        });
+        _i_coeff_binding.watch([this] {
+            reset();
+            _i_coeff = config::shard_local_cfg().fetch_pid_i_coeff();
+        });
+        _d_coeff_binding.watch([this] {
+            reset();
+            _d_coeff = config::shard_local_cfg().fetch_pid_d_coeff();
+        });
+        _target_fetch_sg_util_binding.watch([this] {
+            reset();
+            _target_fetch_sg_util
+              = scaling_factor
+                * config::shard_local_cfg().fetch_pid_target_fetch_sg();
+        });
+        _max_delay_binding.watch([this] {
+            _max_delay = config::shard_local_cfg().fetch_pid_max_debounce();
+        });
+        setup_metrics();
+    }
+
+    std::chrono::milliseconds current_delay(const ss::scheduling_group& sg) {
+        auto sampled_time = ss::sched_clock::now();
+        auto sample = sg.get_stats().runtime;
+        auto total_busy_time = ss::engine().total_busy_time();
+
+        auto dt = sampled_time - _last_sampled_time;
+        if (dt < sample_rate) {
+            return _last_delay;
+        }
+
+        auto busy_de = total_busy_time - _last_total_busy_time;
+        if (busy_de < 0 * duration_unit) {
```

Comment:
Note you can just us `0s` or whatever other unit suffix you want: comparisons work fine between units of different types.